src/ILogger/ApplicationInsightsLogger.cs

```diff
@@ -0,0 +1,203 @@
+        }
+
+        /// <summary>
+        /// Gets or sets the external scope provider.
+        /// </summary>
+        internal IExternalScopeProvider ExternalScopeProvider { get; set; }
+
+        /// <summary>
+        /// Begins a logical operation scope.
+        /// </summary>
+        /// <typeparam name="TState">Current state.</typeparam>
+        /// <param name="state">The identifier for the scope.</param>
+        /// <returns>
+        /// An IDisposable that ends the logical operation scope on dispose.
+        /// </returns>
+        public IDisposable BeginScope<TState>(TState state)
+        {
+            return this.ExternalScopeProvider != null ? this.ExternalScopeProvider.Push(state) : NullScope.Instance;
+        }
+
+        /// <summary>
+        /// Checks if the given <paramref name="logLevel" /> is enabled.
+        /// </summary>
+        /// <param name="logLevel">level to be checked.</param>
+        /// <returns>
+        ///   <c>true</c> if enabled.
+        /// </returns>
+        public bool IsEnabled(LogLevel logLevel)
+        {
+            return this.telemetryClient.IsEnabled();
+        }
+
+        /// <summary>
+        /// Writes a log entry.
+        /// </summary>
+        /// <typeparam name="TState">State being passed along.</typeparam>
+        /// <param name="logLevel">Entry will be written on this level.</param>
+        /// <param name="eventId">Id of the event.</param>
+        /// <param name="state">The entry to be written. Can be also an object.</param>
+        /// <param name="exception">The exception related to this entry.</param>
+        /// <param name="formatter">Function to create a <c>string</c> message of the <paramref name="state" /> and <paramref name="exception" />.</param>
+        public void Log<TState>(LogLevel logLevel, EventId eventId, TState state, Exception exception, Func<TState, Exception, string> formatter)
+        {
+            if (this.IsEnabled(logLevel))
+            {
+                if (exception == null || this.applicationInsightsLoggerOptions?.TrackExceptionsAsExceptionTelemetry == false)
+                {
+                    TraceTelemetry traceTelemetry = new TraceTelemetry(formatter(state, exception), this.GetSeverityLevel(logLevel));
+                    this.PopulateTelemetry(traceTelemetry, state, eventId);
+                    this.telemetryClient.TrackTrace(traceTelemetry);
+                }
+                else
+                {
+                    ExceptionTelemetry exceptionTelemetry = new ExceptionTelemetry(exception)
+                    {
+                        Message = formatter(state, exception),
+                        SeverityLevel = this.GetSeverityLevel(logLevel),
+                    };
+                    exceptionTelemetry.Context.Properties["Exception"] = exception.ToString();
```

Comment:
What is the point of this? Isn't the exception object itself enough?

Reply:
I picked this part of code from @SergeyKanzhelev 's branch. Who in turn IIRC picked from the existing ILogger implementation for ASP.NET Core. Changing this implementation might change the behavior of people using ASP.NET Core once they upgrade. If we are fine with it. I can look into changing this.

Reply:
See https://github.com/Microsoft/ApplicationInsights-aspnetcore/pull/574 for some background info on above code

Reply:
I think @pharring 's point was that why are we double logging the exception object. Thing is ExceptionTelemetry is parsing the stack trace and then pulling information from it and putting it in parsedStack. Adding this property gives you back to full Exception few that you would normally get when IDE (VS) shows it. There are advantages of adding this and disadvantages.

Advantage is you get the full stack that you would have otherwise have seen. However this information is already present in the parsed stack. So we are not really gaining anything very special. Also this is adding a lot of bits over the wire and thus will contribute to user's quota.

Based on the existing non ILogger implementation [here](https://github.com/Microsoft/ApplicationInsights-aspnetcore/blob/2e5aa6c9cf787123990cf56a2505d7137f1704e5/src/Microsoft.ApplicationInsights.AspNetCore/DiagnosticListeners/Implementation/HostingDiagnosticListener.cs#L474). I would recommend we remove this. We can re-add this if people ask for it. This will obviously cause the new logger to have different behavior than the existing ILogger. So I will let AI team decide on it.

Reply:
Just for reference here is what the payload over the wire looks like if we include **Exception** property


```json
{
    "name": "Microsoft.ApplicationInsights.Dev.00000000000000000000000000000000.Exception",
    "time": "2018-12-17T11:27:59.7453991Z",
    "iKey": "00000000-0000-0000-0000-000000000000",
    "tags": {
        "ai.application.ver": "1.0.0.0",
        "ai.cloud.roleInstance": "---",
        "ai.operation.id": "b841d3050a026441b3c8ad9e6e8b7526",
        "ai.operation.parentId": "|b841d3050a026441b3c8ad9e6e8b7526.d7fff5e7_",
        "ai.operation.name": "GET Values/Get",
        "ai.location.ip": "::1",
        "ai.internal.sdkVersion": "il:2.8.1-36930",
        "ai.internal.nodeName": "---"
    },
    "data": {
        "baseType": "ExceptionData",
        "baseData": {
            "ver": 2,
            "exceptions": [{
                "id": 29403036,
                "outerId": 0,
                "typeName": "System.ArgumentNullException",
                "message": "An unhandled exception has occurred while executing the request.",
                "hasFullStack": true,
                "parsedStack": [{
                    "level": 0,
                    "method": "ApplicationInsightsModernLogger.Controllers.ValuesController.Get",
                    "assembly": "ApplicationInsightsModernLogger, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null",
                    "fileName": "D:\\Code\\VSProjects\\ApplicationInsightsModernLogger\\ApplicationInsightsModernLogger\\Controllers\\ValuesController.cs",
                    "line": 27
                }, {
                    "level": 1,
                    "method": "lambda_method",
                    "assembly": "Anonymously Hosted DynamicMethods Assembly, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null",
                    "line": 0
                }, ...... REST OF THE PARSED STACK TRACE
            }],
            "severityLevel": "Error",
            "properties": {
                "CategoryName": "Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddleware",
                "RequestId": "0HLJ42AH8F859:00000001",
                "AspNetCoreEnvironment": "Development",
                "RequestPath": "/api/values",
                "DeveloperMode": "true",
                "{OriginalFormat}": "An unhandled exception has occurred while executing the request.",
                "Exception": "System.ArgumentNullException: Value cannot be null.\r\nParameter name: testParam\r\n   at ApplicationInsightsModernLogger.Controllers.ValuesController.Get() in D:\\Code\\VSProjects\\ApplicationInsightsModernLogger\\ApplicationInsightsModernLogger\\Controllers\\ValuesController.cs:line 27\r\n   at lambda_method(Closure , Object , Object[] )\r\n   at Microsoft.Extensions.Internal.ObjectMethodExecutor.Execute(Object target, Object[] parameters)\r\n   at Microsoft.AspNetCore.Mvc.Internal.ActionMethodExecutor.SyncObjectResultExecutor.Execute(IActionResultTypeMapper mapper, ObjectMethodExecutor executor, Object controller, Object[] arguments)\r\n   at Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker.InvokeActionMethodAsync()\r\n   at Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker.InvokeNextActionFilterAsync()\r\n   at Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker.Rethrow(ActionExecutedContext context)\r\n   at Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)\r\n   at Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker.InvokeInnerFilterAsync()\r\n   at Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.InvokeNextResourceFilter()\r\n   at Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.Rethrow(ResourceExecutedContext context)\r\n   at Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)\r\n   at Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.InvokeFilterPipelineAsync()\r\n   at Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.InvokeAsync()\r\n   at Microsoft.AspNetCore.Builder.RouterMiddleware.Invoke(HttpContext httpContext)\r\n   at Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddleware.Invoke(HttpContext context)",
                "ConnectionId": "0HLJ42AH8F859"
            }
        }
    }
}```

Reply:
Removing this for the initial release.